File: r/R/lookup.R
# Variable lookup over a v3 bundle.
#
# Port of `registream-autolabel/src/registream/autolabel/_lookup.py`
# `lookup()`. Takes a character vector of variable names + optional
# scope/release, returns an `rs_lookup_result` S3 object with:
#   df            : matched metadata rows (one per variable default,
#                   all matching rows when detail = TRUE)
#   missing       : requested names that matched nothing
#   scope_counts  : map of variable_name (lowercase) to distinct scope
#                   tuples the variable appears in

#' @export
rs_lookup <- function(variables,
                      domain    = "scb",
                      lang      = "eng",
                      scope     = NULL,
                      release   = NULL,
                      detail    = FALSE,
                      directory = NULL) {
  if (missing(variables) || is.null(variables) || length(variables) == 0L) {
    stop("rs_lookup() requires a non-empty character vector in `variables`.",
         call. = FALSE)
  }
  variables <- as.character(variables)
  log_and_heartbeat("rs_lookup")

  bundle <- registream::load_bundle(domain, lang, directory = directory)
  filtered <- filter_bundle(bundle, scope = scope, release = release)

  metadata <- filtered$variables
  if (!"variable_name" %in% colnames(metadata) || nrow(metadata) == 0L) {
    return(.rs_lookup_result(
      df              = data.frame(),
      missing         = variables,
      scope_counts    = list()
    ))
  }

  requested_lower <- tolower(variables)
  var_lower <- tolower(as.character(metadata$variable_name))
  matched_mask <- var_lower %in% requested_lower
  matched <- metadata[matched_mask, , drop = FALSE]
  matched_var_lower <- var_lower[matched_mask]

  missing <- variables[!(requested_lower %in% unique(matched_var_lower))]

  scope_counts <- .v3_scope_counts(matched, matched_var_lower, bundle)

  result_df <- if (nrow(matched) == 0L) {
    matched
  } else if (isTRUE(detail)) {
    .v3_attach_scope_tuple(matched, bundle)
  } else {
    .v3_aggregate_default(matched)
  }

  result_df <- .enrich_with_value_labels(result_df, filtered$value_labels)
  rownames(result_df) <- NULL

  .rs_lookup_result(
    df           = result_df,
    missing      = missing,
    scope_counts = scope_counts
  )
}


#' @export
print.rs_lookup_result <- function(x, ...) {
  n <- nrow(x$df)

  if (n == 0L && length(x$missing) == 0L) {
    cat("<rs_lookup_result: empty>\n")
    return(invisible(x))
  }

  if (n > 0L) {
    cat(sprintf("<rs_lookup_result: %d variable%s>\n",
                n, if (n == 1L) "" else "s"))
    display_cols <- intersect(
      c("variable_name", "variable_label", "variable_unit", "variable_type",
        "scope_level_1", "scope_level_2", "release"),
      colnames(x$df)
    )
    if (length(display_cols) > 0L) {
      print(x$df[, display_cols, drop = FALSE], row.names = FALSE)
    }
  } else {
    cat("<rs_lookup_result: 0 variables matched>\n")
  }

  if (length(x$missing) > 0L) {
    cat(sprintf("\nNot found (%d): %s\n",
                length(x$missing),
                paste(x$missing, collapse = ", ")))
  }

  if (length(x$scope_counts) > 0L) {
    counts_vec <- unlist(x$scope_counts, use.names = TRUE)
    multi <- counts_vec[counts_vec > 1L]
    if (length(multi) > 0L) {
      cat(sprintf("\n%d variable%s appear%s in multiple scopes. ",
                  length(multi),
                  if (length(multi) == 1L) "" else "s",
                  if (length(multi) == 1L) "s" else ""))
      cat("Use `detail = TRUE` to see all scope rows.\n")
    }
  }

  invisible(x)
}


# ── Internal helpers ─────────────────────────────────────────────────────────

.rs_lookup_result <- function(df, missing, scope_counts) {
  structure(
    list(
      df           = df,
      missing      = missing,
      scope_counts = scope_counts
    ),
    class = "rs_lookup_result"
  )
}


.v3_aggregate_default <- function(matched) {
  if (nrow(matched) == 0L) return(matched)

  var_name  <- as.character(matched$variable_name)
  var_label <- as.character(matched$variable_label)
  label_key <- paste0(var_name, "\001", var_label)
  label_freq <- stats::ave(seq_len(nrow(matched)), label_key, FUN = length)

  if ("_is_primary" %in% colnames(matched)) {
    ord <- order(var_name,
                 -as.integer(matched$`_is_primary`),
                 -label_freq,
                 method = "radix")
  } else {
    ord <- order(var_name, -label_freq,
                 method = "radix")
  }
  sorted <- matched[ord, , drop = FALSE]
  sorted[!duplicated(sorted$variable_name), , drop = FALSE]
}


.v3_attach_scope_tuple <- function(matched, bundle) {
  depth <- bundle$manifest$scope_depth
  if (depth == 0L ||
      is.null(bundle$scope) ||
      is.null(bundle$release_sets)) {
    return(matched)
  }
  level_cols <- sprintf("scope_level_%d", seq_len(depth))
  scope_cols <- c("scope_id", level_cols,
                  if ("release" %in% colnames(bundle$scope)) "release"
                  else character(0))
  rs <- bundle$release_sets[, c("release_set_id", "scope_id"), drop = FALSE]
  sc <- bundle$scope[, scope_cols, drop = FALSE]
  joined <- merge(matched, rs, by = "release_set_id",
                  all.x = TRUE, sort = FALSE)
  joined <- merge(joined, sc, by = "scope_id",
                  all.x = TRUE, sort = FALSE)
  joined
}


.v3_scope_counts <- function(matched, matched_var_lower, bundle) {
  if (nrow(matched) == 0L ||
      is.null(bundle$scope) ||
      is.null(bundle$release_sets)) {
    return(list())
  }
  depth <- bundle$manifest$scope_depth
  if (depth == 0L) return(list())
  level_cols <- sprintf("scope_level_%d", seq_len(depth))
  rs <- bundle$release_sets[, c("release_set_id", "scope_id"), drop = FALSE]
  sc <- bundle$scope[, c("scope_id", level_cols), drop = FALSE]
  joined <- merge(matched, rs, by = "release_set_id",
                  all.x = TRUE, sort = FALSE)
  joined <- merge(joined, sc, by = "scope_id",
                  all.x = TRUE, sort = FALSE)
  joined$.var_lower <- tolower(as.character(joined$variable_name))

  dedup_key <- do.call(paste,
                       c(list(joined$.var_lower),
                         lapply(level_cols, function(c) as.character(joined[[c]])),
                         list(sep = "\001")))
  uniq <- joined[!duplicated(dedup_key), , drop = FALSE]
  counts <- table(uniq$.var_lower)
  as.list(stats::setNames(as.integer(unname(counts)), names(counts)))
}


.enrich_with_value_labels <- function(df, val_meta) {
  empty <- stats::setNames(integer(0), character(0))

  if (nrow(df) == 0L) {
    df[["value_labels"]] <- list()
    return(df)
  }

  # Parse only the value-label rows actually referenced by `df`. The full
  # SCB-English `value_labels` table is ~12k rows and parsing every row
  # takes ~60s; a single-variable lookup typically references one lid.
  cache <- new.env(parent = emptyenv())
  needed_lids <- if ("value_label_id" %in% colnames(df)) {
    raw_lids <- df[["value_label_id"]]
    raw_lids <- raw_lids[!is.na(raw_lids)]
    unique(trimws(as.character(raw_lids)))
  } else character(0)
  needed_lids <- needed_lids[nzchar(needed_lids)]

  if (length(needed_lids) > 0L &&
      !is.null(val_meta) &&
      "value_label_id" %in% colnames(val_meta) &&
      "value_labels_stata" %in% colnames(val_meta)) {
    val_lid <- trimws(as.character(val_meta[["value_label_id"]]))
    keep <- val_lid %in% needed_lids
    val_meta_sub <- val_meta[keep, , drop = FALSE]
    for (i in seq_len(nrow(val_meta_sub))) {
      lid <- val_meta_sub[["value_label_id"]][[i]]
      if (is.na(lid)) next
      lid_str <- trimws(as.character(lid))
      if (!nzchar(lid_str)) next
      if (exists(lid_str, envir = cache, inherits = FALSE)) next
      raw <- val_meta_sub[["value_labels_stata"]][[i]]
      parsed <- registream::parse_value_labels_stata(raw)
      if (length(parsed) > 0L) {
        assign(lid_str, parsed, envir = cache)
      }
    }
  }

  lookup_one <- function(lid) {
    if (is.na(lid)) return(empty)
    lid_str <- trimws(as.character(lid))
    if (!nzchar(lid_str)) return(empty)
    if (!exists(lid_str, envir = cache, inherits = FALSE)) return(empty)
    get(lid_str, envir = cache, inherits = FALSE)
  }

  if ("value_label_id" %in% colnames(df)) {
    df[["value_labels"]] <- lapply(df[["value_label_id"]], lookup_one)
  } else {
    df[["value_labels"]] <- replicate(nrow(df), empty, simplify = FALSE)
  }

  df
}

File: python/registream-autolabel/src/registream/autolabel/_lookup.py
"""Variable lookup over a loaded :class:`~registream.metadata.LabelBundle`.

Given one or more variable names, return a :class:`LookupResult`
containing:

- the matched metadata rows (one per variable in default mode, all
  matching rows in detail mode);
- the list of variable names that were requested but not found;
- ``scope_counts``: the number of distinct scope tuples per matched
  variable: useful for surfacing "appears in N scopes" in the UI.

The caller passes a pre-filtered :class:`~registream.autolabel._types.FilteredBundle`
(scope / release filtering already applied). This mirrors Stata's
flow where ``autolabel lookup`` operates on the already-filtered
metadata.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import pandas as pd

from registream.autolabel._labels import (
    VALUE_LABEL_ID_COLUMN,
    VALUE_LABELS_COLUMN,
    VARIABLE_LABEL_COLUMN,
    VARIABLE_NAME_COLUMN,
    parse_value_labels_stata,
)
from registream.autolabel._types import FilteredBundle, LabelBundle

__all__ = ["LookupResult", "lookup"]


_DISPLAY_COLUMNS: tuple[str, ...] = (
    "variable_name",
    "variable_label",
    "variable_type",
    "variable_unit",
    "value_labels",
    "scope_level_1",
    "scope_level_2",
    "release",
)


@dataclass
class LookupResult:
    """Result of a :func:`lookup` call.

    :attr:`df` carries the matched metadata rows (enriched with a
    ``value_labels`` dict column parsed from ``value_labels_stata``).
    :attr:`scope_counts` maps lowercased ``variable_name`` to the
    number of distinct scope tuples the variable appears in within the
    (filtered) bundle.

    The object renders as a DataFrame in Jupyter (``_repr_html_``) and
    as a compact table in a plain REPL (``__repr__``); the wrapping
    is for attribute access (``.missing``, ``.scope_counts``) without
    forcing callers to reach into ``.df``.
    """

    df: pd.DataFrame
    missing: list[str]
    scope_counts: dict[str, int] = field(default_factory=dict)

    def _display_df(self) -> pd.DataFrame:
        """Curated column subset for rendering.

        The raw ``df`` holds a dozen columns (definition, source, FK
        ids, …); for display we keep the fields users actually read so
        the table fits typical Jupyter / terminal widths. Call
        ``.df`` directly for the full row.
        """
        if self.df.empty:
            return self.df
        cols = [c for c in _DISPLAY_COLUMNS if c in self.df.columns]
        return self.df[cols] if cols else self.df

    def __repr__(self) -> str:
        if self.df.empty:
            body = "(no matches)"
        else:
            body = repr(self._display_df())
        missing = f"\nMissing: {self.missing}" if self.missing else ""
        return body + missing

    def _repr_html_(self) -> str:
        if self.df.empty:
            body = "<p><i>no matches</i></p>"
        else:
            html = getattr(self._display_df(), "_repr_html_", None)
            body = html() if callable(html) else repr(self._display_df())
        if self.missing:
            body += f"<p><b>Missing:</b> {', '.join(self.missing)}</p>"
        return body


def lookup(
    variables: str | list[str],
    bundle: LabelBundle,
    filtered: FilteredBundle,
    *,
    detail: bool = False,
) -> LookupResult:
    """Look up metadata for ``variables`` in the filtered bundle."""
    if isinstance(variables, str):
        variables = [variables]

    requested_lower = [v.lower() for v in variables]
    requested_pairs = list(zip(variables, requested_lower, strict=False))

    metadata = filtered.variables
    if VARIABLE_NAME_COLUMN not in metadata.columns or metadata.empty:
        return LookupResult(df=pd.DataFrame(), missing=list(variables))

    metadata = metadata.copy()
    metadata["_var_lower"] = metadata[VARIABLE_NAME_COLUMN].astype(str).str.lower()
    matched = metadata[metadata["_var_lower"].isin(requested_lower)]

    matched_lower = set(matched["_var_lower"]) if not matched.empty else set()
    missing = [orig for orig, low in requested_pairs if low not in matched_lower]

    scope_counts = _scope_counts(matched, bundle)

    if matched.empty:
        result_df = matched
    elif detail:
        result_df = _attach_scope_tuple(matched, bundle)
    else:
        result_df = _aggregate_default(matched)

    result_df = _enrich_with_value_labels(result_df, filtered.value_labels)
    if "_var_lower" in result_df.columns:
        result_df = result_df.drop(columns=["_var_lower"])

    return LookupResult(
        df=result_df.reset_index(drop=True),
        missing=missing,
        scope_counts=scope_counts,
    )


def _aggregate_default(matched: pd.DataFrame) -> pd.DataFrame:
    """Pick the most-common label per variable and return one row each."""
    df = matched.copy()
    df["_label_freq"] = df.groupby(
        [VARIABLE_NAME_COLUMN, VARIABLE_LABEL_COLUMN]
    )[VARIABLE_NAME_COLUMN].transform("size")
    sort_cols = [VARIABLE_NAME_COLUMN, "_label_freq"]
    ascending = [True, False]
    if "_is_primary" in df.columns:
        sort_cols.insert(1, "_is_primary")
        ascending.insert(1, False)
    sorted_df = df.sort_values(by=sort_cols, ascending=ascending, kind="stable")
    deduped = sorted_df.drop_duplicates(subset=[VARIABLE_NAME_COLUMN], keep="first")
    return deduped.drop(columns=["_label_freq"])


def _attach_scope_tuple(matched: pd.DataFrame, bundle: LabelBundle) -> pd.DataFrame:
    """Add ``scope_level_1..N`` and ``release`` columns via the FK chain."""
    depth = bundle.manifest.scope_depth
    if depth == 0 or bundle.scope is None or bundle.release_sets is None:
        return matched

    level_cols = [f"scope_level_{i}" for i in range(1, depth + 1)]
    scope_cols = ["scope_id", *level_cols, "release"] if "release" in bundle.scope.columns else ["scope_id", *level_cols]
    joined = matched.merge(bundle.release_sets, on="release_set_id", how="left").merge(
        bundle.scope[scope_cols], on="scope_id", how="left"
    )
    return joined


def _scope_counts(matched: pd.DataFrame, bundle: LabelBundle) -> dict[str, int]:
    """Count the distinct scope tuples per (lowered) ``variable_name``."""
    if matched.empty or bundle.scope is None or bundle.release_sets is None:
        return {}
    depth = bundle.manifest.scope_depth
    if depth == 0:
        return {}
    level_cols = [f"scope_level_{i}" for i in range(1, depth + 1)]
    joined = matched.merge(bundle.release_sets, on="release_set_id", how="left").merge(
        bundle.scope[["scope_id", *level_cols]], on="scope_id", how="left"
    )
    grouped = (
        joined.drop_duplicates(subset=["_var_lower", *level_cols])
        .groupby("_var_lower")
        .size()
    )
    return {str(k): int(v) for k, v in grouped.to_dict().items()}


def _enrich_with_value_labels(
    df: pd.DataFrame,
    value_labels: pd.DataFrame | None,
) -> pd.DataFrame:
    df = df.copy()
    if df.empty:
        df["value_labels"] = pd.Series([], dtype=object)
        return df

    # Parse only the value-label rows actually referenced by `df`. The full
    # SCB-English value_labels table is ~12k rows; eager parsing scales O(n)
    # in the bundle even for a single-variable lookup.
    cache: dict[str, dict] = {}
    needed_lids: set[str] = set()
    if VALUE_LABEL_ID_COLUMN in df.columns:
        for lid in df[VALUE_LABEL_ID_COLUMN].dropna():
            key = str(lid).strip()
            if key:
                needed_lids.add(key)

    if (
        needed_lids
        and value_labels is not None
        and VALUE_LABEL_ID_COLUMN in value_labels.columns
        and VALUE_LABELS_COLUMN in value_labels.columns
    ):
        lid_series = value_labels[VALUE_LABEL_ID_COLUMN].astype(str).str.strip()
        sub = value_labels[lid_series.isin(needed_lids)]
        for _, row in sub.iterrows():
            label_id = row[VALUE_LABEL_ID_COLUMN]
            if pd.isna(label_id):
                continue
            label_id_str = str(label_id).strip()
            if not label_id_str or label_id_str in cache:
                continue
            parsed = parse_value_labels_stata(row[VALUE_LABELS_COLUMN])
            if parsed:
                cache[label_id_str] = parsed

    def _one(lid: object) -> dict:
        if lid is None or (isinstance(lid, float) and pd.isna(lid)):
            return {}
        key = str(lid).strip()
        if not key:
            return {}
        return dict(cache.get(key, {}))

    if VALUE_LABEL_ID_COLUMN in df.columns:
        df["value_labels"] = df[VALUE_LABEL_ID_COLUMN].apply(_one)
    else:
        df["value_labels"] = [{} for _ in range(len(df))]
    return df
